Player: Auto select current video in side drawer

mesalilac opened the pull request:
Adds a button in the `SideDrawer` of the video player to allow the user to jump to the currently playing video.

## Notes
- I used the compose icon because i couldn't find a better icon!

## Demo
![Pi7_GIF_CMP](https://github.com/user-attachments/assets/6bdafe7c-6532-4ecf-b689-c60d5d3bc769)

mesalilac committed:
Pass `currentlyPlayingVideoID` into `SideDrawer`

mesalilac committed:
Add button `jump-to-currently-playing-video`

mesalilac committed:
Add `jump-to-currently-playing-btn` class

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Stremio/stremio-web?pullRequest=936) <br/>All committers have signed the CLA.

tymmesyde commented:
Hi, i think this should be the default behavior, so no need for a button

mesalilac commented:
> Hi, i think this should be the default behavior, so no need for a button

Good idea :+1:

mesalilac committed:
automatically jump to video on mount, remove button

mesalilac commented:
I'm facing a problem right now, the [`useEffect`](https://github.com/Stremio/stremio-web/pull/936/files#diff-083c71782a3716e7f1cb143b6c9bdd5a7ae217d5c7d5c49dca67ca29b22214a7R81) run takes a long time to `scrollIntoView` when the video list is long.
does anyone have a better way to do it?

> [!NOTE]
> The reason i have a `setTimeout` delay inside the `useEffect` is because `scrollIntoView` moves the sidebar elements if the sidebar transition is not complete.

kKaskak commented:
> I'm facing a problem right now, the [`useEffect`](https://github.com/Stremio/stremio-web/pull/936/files#diff-083c71782a3716e7f1cb143b6c9bdd5a7ae217d5c7d5c49dca67ca29b22214a7R81) run takes a long time to `scrollIntoView` when the video list is long. does anyone have a better way to do it?
> 
> Note
> 
> The reason i have a `setTimeout` delay inside the `useEffect` is because `scrollIntoView` moves the sidebar elements if the sidebar transition is not complete.

You can use `onTransitionEnd` to track when the transition ends and then fire the scroll. 

### TODO
- [x] use `onTransitionEnd` to eliminate the need for `setTimeout`
- [x] use `forwardRef` instead of creating a container wrapper for the `Video` component
- [x] clear prop naming

Thank you for the contribution

mesalilac committed:
use `onTransitionEnded` and remove `setTimeout`

mesalilac committed:
remove container and use `forwardRef`

mesalilac committed:
clear up naming

mesalilac committed:
fix: pass correct prop to `SideDrawer`

mesalilac committed:
fix: SideDrawer glitched element movement.
Even when using `onTransitionEnd`, the `scrollIntoView` problem persists.
The solution that i found is to use `requestAnimationFrame` two times,
turns out that's enough time after the component mounts to fix the problem.

kKaskak commented on src/routes/Player/SideDrawer/SideDrawer.tsx:
```diff
@@ -75,8 +76,21 @@ const SideDrawer = memo(forwardRef<HTMLDivElement, Props>(({ seriesInfo, classNa
+        const { current } = selectedVideoRef;
+        if (!current) return;
+
+        requestAnimationFrame(() => {
```
You can use `onTransitionEnd` from the `Transition` component to avoid using workarounds. Invoke the scroll after it ends.

mesalilac replied:
I couldn't find `onTransitionEnded` anywhere! 
could you provide more info

kKaskak replied:
https://github.com/Stremio/stremio-web/blob/development/src/components/Transition/Transition.tsx#L48

Misspelled the function name sorry

mesalilac replied:
when i try `onTransitionEnded` insted of `onTransitionEnd`
```tsx
        <div ref={ref} className={classNames(styles['side-drawer'], className)} onTransitionEnded={jumpToPlayingNow} onMouseDown={onMouseDown}>

```
Throws: "Warning: Unknown event handler property `onTransitionEnded`. It will be ignored."

mesalilac replied:
> https://github.com/Stremio/stremio-web/blob/development/src/components/Transition/Transition.tsx#L48
> 
> Misspelled the function name sorry

Sorry i don't know what you mean
do i use the `ref` passed to the `SideDrawer` to do this in side a useEffect?
```tsx
        ref?.addEventListener('transitionend', onTransitionEnd);
        return () => ref?.removeEventListener('transitionend', onTransitionEnd);
```
could you show me a code example

kKaskak replied:
> > https://github.com/Stremio/stremio-web/blob/development/src/components/Transition/Transition.tsx#L48
> > Misspelled the function name sorry
> 
> Sorry i don't know what you mean do i use the `ref` passed to the `SideDrawer` to do this in side a useEffect?
> 
> ```tsx
>         ref?.addEventListener('transitionend', onTransitionEnd);
>         return () => ref?.removeEventListener('transitionend', onTransitionEnd);
> ```
> 
> could you show me a code example

I was trying to point you in the correct direction so you can think about the solution yourself. The idea is to track when the transition on the `Transition` component ends and then pass it as a prop to the `SideDrawer` and invoke the scroll when its true (meaning the transition ended). You don't need to add `onTransitionEnded` to the `SideDrawer`. Try to think how you can track the transition timing and pass it to the `SideDrawer` as a `boolean` value to then do something like this: 

```
    useEffect(() => {
        if (transitionEnded) {
            selectedVideoRef.current?.scrollIntoView({ behavior: 'smooth', block: 'center' });
        }
    }, [transitionEnded]);
```

mesalilac committed:
add a callback function prop `onTransitionEnd`

mesalilac committed:
pass `sideDrawerTransitionEnded` into `SideDrawer`

mesalilac committed:
only scroll if `transitionEnded` is true

kKaskak committed:
Merge branch 'development' into jump-to-currently-playing

kKaskak committed:
chore: fix builds

kKaskak committed:
refactor(SideDrawer): simplify refs handing

kKaskak commented:
fyi you can look up the code to see how much cleaner the solution i tried to explain to you is. 
1. Track when the transition ends with a state or variable inside `Transition` component.
2. Pass it to `SideDrawer` component
3. Invoke scroll.

That's it 🎊 

Please test it as well 👍🏼

mesalilac commented:
a3c895dfc6589e6b0083061885cbcd523e2ea181. Thank you, that was the obvious change that i should have made

mesalilac commented:
I have tested in "Movies", "Series", and "Channels". it worked perfectly with no errors.

kKaskak commented:
Great
